src/goto-programs/remove_complex.cpp

```diff
@@ -131,7 +129,8 @@ static void remove_complex(exprt &expr)
     if(expr.id()==ID_plus || expr.id()==ID_minus ||
        expr.id()==ID_mult || expr.id()==ID_div)
     {
-      assert(expr.operands().size()==2);
+      DATA_INVARIANT(
+        expr.operands().size() == 2, "binary operators have 2 arguments");
```

Comment:
But plus and mult are not binary:

https://github.com/diffblue/cbmc/blob/develop/src/util/std_expr.h#L963
https://github.com/diffblue/cbmc/blob/develop/src/util/std_expr.h#L1080

I realise it may be beyond the remit of this PR to fix this but this needs to be a PRECONDITION stating clearly that _in_this_context_ they must have 2 arguments.

Reply:
That's not strictly correct.  ID_plus, ID_mult and the other associative operators below are really genuinely, always not binary operators in the sense they do not inherit from binary_exprt.  However, in common operation, they often only have two operands.

There used to be some confusion over whether they were binary or n-arty so there is some code that is written with an implicit assumption that they are binary.  This and remove_vector are examples.  The ambiguity was resolved in favour of them becoming n-ary so this code is wrong.

It is indeed a DATA_INVARIANT that binary operators have two arguments but these are not binary operators so I would recommend a PRECONDITION instead, along with a FIXME saying it needs to handle the case of n-ary plus and mult.  Similar should be applied below.